File: version4j/src/main/java/com/only4/algorithm/version4j/leetcode/LevelOrder.java
package com.only4.algorithm.version4j.leetcode;

import com.only4.algorithm.version4j.extra.TreeNode;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class LevelOrder {
    public List<List<Integer>> levelOrder(TreeNode root) {
        if (root == null) return Collections.emptyList();

        ArrayDeque<TreeNode> queue = new ArrayDeque<>();
        List<List<Integer>> result = new ArrayList<>();

        queue.add(root);

        while (!queue.isEmpty()) {
            int currentLevelSize = queue.size();
            List<Integer> currentLevel = new ArrayList<>();

            while (currentLevelSize-- > 0) {
                TreeNode currentNode = queue.pollFirst();
                currentLevel.add(currentNode.val);

                if (currentNode.left != null) queue.addLast(currentNode.left);
                if (currentNode.right != null) queue.addLast(currentNode.right);
            }
            result.add(currentLevel);
        }
        return result;
    }
}

File: version4k/src/main/kotlin/com/only4/algorithm/version4k/leetcode/levelOrder.kt
package com.only4.algorithm.version4k.leetcode

import com.only4.algorithm.version4k.extra.TreeNode

fun levelOrder(root: TreeNode?): List<List<Int>> {
    if (root == null) return emptyList()

    val queue = ArrayDeque<TreeNode>().apply { addLast(root) }
    val result = mutableListOf<List<Int>>()

    while (queue.isNotEmpty()) {
        val currentLevel = mutableListOf<Int>()
        val currentLevelSize = queue.size

        repeat(currentLevelSize) {
            val currentNode = queue.removeFirst()
            currentLevel.add(currentNode.`val`)

            currentNode.left?.let { queue.addLast(it) }
            currentNode.right?.let { queue.addLast(it) }
        }
        result.add(currentLevel)
    }

    return result
}
